### PR TITLE
Use the Identity_Crisis package

### PR DESCRIPTION
**Pull request by briancolinger**
Update Jetpack to use the Identity_Crisis package. Included is migration of Sync Functions home_url, site_url and dependent functions to the Connections package so that we have the following package dependencies

Sync -> IDC + Connections
IDC -> Connections

Regression Testing of IDC is needed to  merge this PR.

#### Changes proposed in this Pull Request:

* Replaces the usage of Jetpack_IDC with Identity_Crisis.
* Maintains Jetpack_IDC class for backward compatibility during upgrades.
* Migrate Sync/Function url functions into Connection/Utils
* Remove Identity Crisis dependencies on Sync.
* Update Connection to use normalized site_url and home_url during registration flow.

#### Does this pull request change what data or activity we track or use?

NO

#### Testing instructions:

Regression Testing
1) Verify that new site registration have home_url and site_url as expected on the WP.com Cache (E2E flows cover)
2) Verify that Identity Crisis flows are fully functional 
 -- Duplicate your JP connected site to a new site (update options to new url except leave jetpack_private_options untouched)
 -- Perform an edit to a post (triggers sync)
 -- Visit the Jetpack Dashboard
 -- Safe Mode will be enabled and IDC options presented.
 -- Verify that choosing this is a new site, doesn't modify the home_url or site_url of the existing cache site and instead the new site has its own cache site created and jetpack_private_options updated. 
 -- Verify that choosing this is the site has stats migrated and home_url and site_url of the existing cache site is updated.
3) Modify a site's site_url and home_url. Ensure these are synced to the WP.com Cache (unit tests cover)

Note: Quickest way to trigger IDC flows is to copy jetpack_private_options from Site A to Site B then trigger a content change on Site B (edit post, etc)

### CODE REVIEW

**Comment by briancolinger** [comment]
I may have made a mistake in the commit... It looks like the identity-crisis package got included by mistake...
Should I remove that from this PR or will it get handled in the merge?

**Comment by briancolinger** [comment]
I've rebased this branch and tidied things up a bit.
Ready for review.

**Comment by kraftbj** [comment]
Marking as blocked until 18869 is in.

**Review by kraftbj — DISMISSED** [review]
Please remove the deleted files from the lining list per the annotation at the end of https://github.com/Automattic/jetpack/pull/19126/files

File: projects/plugins/jetpack/class.jetpack.php
Comment (mdbitz) [inline projects/plugins/jetpack/class.jetpack.php:13]: sync_idc_optin and get_sync_error_idc_optin need to call the IDC packages version.

Comment (mdbitz) [inline projects/plugins/jetpack/class.jetpack.php:13]: Sync Package needs to be updated to call IDC version of sync_idc_optin and get_sync_error_idc_optin

Comment (mdbitz) [reply projects/plugins/jetpack/class.jetpack.php:13]: mark these functions as deprecated.
Update all references in Jetpack to call IDC versions.

File: projects/packages/sync/src/class-main.php
Comment (jeherve) [inline projects/packages/sync/src/class-main.php:28]: Is there no need to import `Automattic\Jetpack\Identity_Crisis` here?

File: projects/plugins/jetpack/_inc/class.jetpack-provision.php
Comment (jeherve) [inline projects/plugins/jetpack/_inc/class.jetpack-provision.php:39]: Same as above, would we need to import `Automattic\Jetpack\Identity_Crisis` here?

File: projects/plugins/jetpack/composer.json
Comment (jeherve) [inline projects/plugins/jetpack/composer.json:41]: Do you think you could keep that list ordered alphabetically, and use a version other than `@dev`? Maybe `0.1.x-dev` is what's appropriate here until we do a first release?

File: projects/plugins/jetpack/class.jetpack-idc.php
Comment (jeherve) [inline projects/plugins/jetpack/class.jetpack-idc.php:1]: I would recommend keeping the file around for a few releases, to avoid any Fatals during the plugin update for folks with aggressive cache setups.
See this example: 
https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/class.jetpack-affiliate.php#L1-L5

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/composer.json
Comment (kraftbj) [reply projects/plugins/jetpack/composer.json:41]: Also watch the, I assume, unintentional downgrade of the twitter text package.

File: projects/plugins/jetpack/class.jetpack-idc.php
Comment (mdbitz) [reply projects/plugins/jetpack/class.jetpack-idc.php:1]: will do thanks for looking out for and the great feedback.

File: projects/plugins/jetpack/tests/php/general/test_class.jetpack.php
Comment (kbrown9) [inline projects/plugins/jetpack/tests/php/general/test_class.jetpack.php:374]: It looks like some (possibly all) of these changed tests should be in the identity-crisis package. I don't think this is a blocker. It's fine with me if we open a new PR for moving the tests.

File: projects/plugins/jetpack/_inc/class.jetpack-provision.php
Comment (mdbitz) [reply projects/plugins/jetpack/_inc/class.jetpack-provision.php:39]: :thumbsup:

File: projects/packages/identity-crisis/src/class-functions.php
Comment (kbrown9) [inline projects/packages/identity-crisis/src/class-functions.php:5]:
```suggestion
 * @package automattic/jetpack-identity-crisis
```

File: projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php
Comment (kbrown9) [inline projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php:8]:
```suggestion
use Automattic\Jetpack\Identity_Crisis\Functions as Identity_Functions;
```

**Comment by kbrown9** [comment]
I've done some testing, and I noticed a few things that might need to be changed in the `Identity_Crisis` class:

* Should this `require_once` be a `require`, since we're obtaining data from the file? https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L490

* I think this `jetpack-idc.css` file is in the `build/css` directory: https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L549

* I think the `plugin_dir_url` call might be unnecessary here. In my tests, the path isn't correct. https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L511

* I'm not sure that we want to call `esc_html` on things like `$this->get_first_step_header_lead()` since that method returns html. https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L620

* `Manager::is_active` has been deprecated. Based on the changes in [PR #19276](https://github.com/Automattic/jetpack/pull/19276/files#diff-028fe0b6a09c6ed2b0a61f21d8db4d21cf8ca35c8904f94f78503619455fe31fL6188), I think we need to use something like `Manager::is_connected` instead. https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L225

File: projects/packages/connection/src/class-utils.php
Comment (mdbitz) [reply projects/packages/connection/src/class-utils.php:22]: Migrated to Urls class in Connection Package.

**Comment by mdbitz** [comment]
Sync/Functions now reside in Connection/Urls
Updated usage of is_active to is_connected
PHPCS ignores on existing echos that are escaped in functions
Update path on files that were incorrect

File: projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php
Comment (kbrown9) [inline projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php:765]: I think we need to add `use Automattic\Jetpack\Connection\Urls;` to this file.

Comment (mdbitz) [reply projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php:765]: good catch no namespace on this class :)

**Comment by kbrown9** [comment]
I think these two `require_once` expressions in the `Identity_Crisis` class should use `require`:

* https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L490
* https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L506

The `build/index.asset.php` file overrides the `require` return value, and we use that return value to set the value of a variable. I don’t think it makes sense to use `require_once` here since we want to evaluate that file every time the methods are called, not just once.

This warning is currently being generated:

```
PHP Warning:  Cannot use a scalar value as an array in …/identity-crisis/src/class-identity-crisis.php on line 507
```

**Comment by kbrown9** [comment]
I’ve done some basic testing of this PR, and it looks good so far in my tests (except for the `require_once` issue described above). More extensive testing is still needed, but here are the details of how I’ve been testing:

* I caused an IDC on my test site using this snippet:
```
add_filter( 'jetpack_sync_site_url', function ( $value ) {
	if( defined('XMLRPC_REQUEST') && XMLRPC_REQUEST ) {
		return $value;
	}
	return 'http://example.org/site';
} );
```
* I tested with these plugins, and the IDC UI was displayed as expected:

  * Jetpack alone.
  * Jetpack and the client-example plugin, with client-example using the connection, sync, and identity-crisis packages.
  * The client-example plugin alone using the connection, sync, and identity-crisis packages.


To test with the client-example plugin, I used [this branch](https://github.com/Automattic/client-example/tree/update/test_sync_idc)  in my Jetpack Docker development environment. Note that client-example’s `composer.json` file uses the monorepo packages paths, allowing client-example to use the in-development sync and identity-crisis packages.

**Comment by mdbitz** [comment]
I've tested this in detail using the Jetpack Beta Builder.

Setup 2 test sites siteA and siteB that had Jetpack and Jetpack Beta plugins installed.

Site A - connected to WP.com
Site B -> duplicated db of SiteA (with jetpack_private_options that had tokens)
Site B -> updated db options to have correct URL
Site B -> edit post
Site B -> visit Jetpack Dashboard
Site B -> verified Safe Mode / IDC prompts are displayed
SIte B -> Verified that stating this was the site caused the SiteA cache site to be updated with new home/site url
WP.com Network Admin -> reset home_url and site_url to Site A values
Site B -> edit post
Site B -> Visit Jetpack Dashboard
Site B -> Verified Safe Mode / IDC prompts are displayed
Site B -> New site causes a new cache site to be created after going through the connection process.
WP.com Network Admin -> verify cache of Site A not modified

File: projects/plugins/jetpack/class.jetpack-idc.php
Comment (kraftbj) [reply projects/plugins/jetpack/class.jetpack-idc.php:1]: Are we going to update the functions within to call the canonical version in this PR or a follow-up?

Comment (mdbitz) [reply projects/plugins/jetpack/class.jetpack-idc.php:1]: @kraftbj this file shouldn't be loaded/used. As Jeremy outline it is only here to avoid fatals, I don't see a need to have these functions reference the package, but I may be overlooking use cases.

Comment (mdbitz) [reply projects/plugins/jetpack/class.jetpack-idc.php:1]: Spent a little time thinking this through and as discussed in slack there are no known references to the Jetpack_IDC class. The intent here is to ensure no fatals during upgrade processes which is why the docblock is added. If another plugin references Jetpack_IDC directly it will fail, the only reason this would be available was if they include the file directly which would potentially cause a conflict if both the new and old IDC classes tried to add the same ui elements.

As such I'm going to delete the contents of the file in full.

File: projects/packages/connection/changelog/update-use-identity-crisis-package
Comment (anomiex) [inline projects/packages/connection/changelog/update-use-identity-crisis-package:4]:
```suggestion
Significance: minor
Type: added

Added Urls class, migrated from Sync Functions.
```

File: projects/packages/identity-crisis/changelog/update-use-identity-crisis-package
Comment (anomiex) [inline projects/packages/identity-crisis/changelog/update-use-identity-crisis-package:4]:
```suggestion
Significance: minor
Type: changed

Use Connection/Urls for home_url and site_url functions migrated from Sync.
```

File: projects/packages/status/src/class-status.php
Comment (anomiex) [inline projects/packages/status/src/class-status.php:243]: Using `::class` can make it easier for tools to recognize it as a class reference.
```suggestion
		if ( ! $is_staging && method_exists( \Automattic\Jetpack\Identity_Crisis::class, 'validate_sync_error_idc_option' ) && \Automattic\Jetpack\Identity_Crisis::validate_sync_error_idc_option() ) {
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:333]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::get_raw_or_filtered_url' );
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:343]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::home_url' );
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:353]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::site_url' );
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:363]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::main_network_site_url' );
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:398]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::get_protocol_normalized_url' );
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:410]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::get_raw_url' );
```

File: projects/packages/sync/src/class-functions.php
Comment (anomiex) [inline projects/packages/sync/src/class-functions.php:422]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Connection\\Urls::normalize_www_in_url' );
```

File: projects/packages/sync/changelog/update-use-identity-crisis-package
Comment (anomiex) [inline projects/packages/sync/changelog/update-use-identity-crisis-package:4]:
```suggestion
Significance: minor
Type: deprecated

Deprecated URL methods in `Automattic\Jetpack\Sync\Functions` in favor of `Automattic\Jetpack\Connection\Urls`.
```

File: projects/plugins/jetpack/class.jetpack.php
Comment (anomiex) [inline projects/plugins/jetpack/class.jetpack.php:6246]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Identity_Crisis::validate_sync_error_idc_option' );
```

File: projects/plugins/jetpack/class.jetpack.php
Comment (anomiex) [inline projects/plugins/jetpack/class.jetpack.php:6281]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Identity_Crisis::get_sync_error_idc_option' );
```

File: projects/plugins/jetpack/class.jetpack.php
Comment (anomiex) [inline projects/plugins/jetpack/class.jetpack.php:6293]:
```suggestion
		_deprecated_function( __METHOD__, 'jetpack-9.8', '\\Automattic\\Jetpack\\Identity_Crisis::sync_idc_optin' );
```

**Review by anomiex — COMMENTED** [review]
Took a quick look through, mostly with an eye towards monorepo structure.

Didn't see anything major. Mostly I spotted that you used forward slash rather than backslash in namespace names in calls to `_deprecated_function()`, and suggested some changes to changelog entries to focus more on the important changes.

File: projects/packages/status/src/class-status.php
Comment (leogermani) [reply projects/packages/status/src/class-status.php:243]: @anomiex but wouldn't this trigger an error if the class does not exist? which is precisely what we are testing there?

Comment (leogermani) [reply projects/packages/status/src/class-status.php:243]: well, not "precisely", but "one of the things"

Comment (mdbitz) [reply projects/packages/status/src/class-status.php:243]: Yes this change would cause an issue and isn't something we want to make.

**Comment by mdbitz** [comment]
Did another round of testing after merging @anomiex Brad's suggestions (Thank you as always)

@kbrown9 or @leogermani do you want to give this a final last test / approval?

**Review by leogermani — DISMISSED** [review]

File: projects/packages/status/src/class-status.php
Comment (anomiex) [reply projects/packages/status/src/class-status.php:243]: > but wouldn't this trigger an error if the class does not exist?

No, it wouldn't. The `Foo::class` syntax is syntactic sugar for the corresponding string, but has the advantage that it's written out like any other class reference and follows `use` directives and the like as other class references do.

See https://3v4l.org/kEsCg for examples.

**Comment by mdbitz** [comment]
Hopefully this should make all test pass 🤞 ⚔️

**Review by mdbitz — APPROVED** [review]
Regression Tested after all the merges. IDC panel displayed and options functional.

**Pull request merged**
